lib/src/main/kotlin/bstrees/AbstractBSTrees.kt

```diff
@@ -7,24 +7,62 @@ abstract class BinarySearchTree<T : Comparable<T>, NodeType : TreeNode<T, NodeTy
     protected var treeRoot: NodeType? = null
 
     fun search(data: T): T? = searchNode(data)?.data
+
+    /**
+     * Searches for node and returns it as a result (or null).
+     */
     protected fun searchNode(data: T): NodeType? {
-        // searches for node and returns it as a result (or null)
-        // aside from public 'search' method
-        // it might be useful for deletion implementations to find the node to delete
-        // that's why we're keeping it visible to inherited classes
-        TODO("Not yet implemented")
+        var currentNode = treeRoot
+        while (currentNode != null) {
+            val res = data.compareTo(currentNode.data)
+            currentNode = when {
+                res < 0 -> currentNode.left
+                res > 0 -> currentNode.right
+                else -> return currentNode
+            }
+        }
+        return null
     }
 
     open fun insert(data: T) {
         insertNode(data)
     }
 
     protected abstract fun createNewNode(data: T): NodeType
+
+    /**
+     * Does simple insert and returns inserted node
+     * Uses [createNewNode] to create new node to insert
+     */
     protected fun insertNode(data: T): NodeType {
-        // does simple insert and returns inserted node
-        // uses 'createNewNode' to create new node to insert
-        // will be used in different implementations of trees as part of the insert process
-        TODO("Not yet implemented")
+        val createdNode = createNewNode(data)
+
+        if (treeRoot == null) {
+            treeRoot = createdNode
+            return createdNode
+        }
+
+        var currentNode = treeRoot!!
+        while (true) {
+            val res = data.compareTo(currentNode.data)
+            if (res < 0) {
+                if (currentNode.left == null) {
+                    currentNode.left = createdNode
+                    createdNode.parent = currentNode
+                    return createdNode
+                }
+                currentNode = currentNode.left!!
+            }
+            else if (res > 0) {
+                if (currentNode.right == null) {
+                    currentNode.right = createdNode
+                    createdNode.parent = currentNode
+                    return createdNode
+                }
+                currentNode = currentNode.right!!
+            }
+            else return currentNode
```

Comment:
That's not how you should handle equality case. We discussed the proper solution earlier: 
```suggestion
            else {
                currentNode.data = data
                return currentNode
            }
```